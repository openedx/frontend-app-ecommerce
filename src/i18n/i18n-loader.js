/**
 * For each locale we want to support, react-intl needs 1) the locale-data, which includes
 * information about how to format numbers, handle plurals, etc., and 2) the translations, as an
 * object holding message id / translated string pairs.  A locale string and the messages object are
 * passed into the IntlProvider element that wraps your element hierarchy.
 *
 * Note that react-intl has no way of checking if the translations you give it actually have
 * anything to do with the locale you pass it; it will happily use whatever messages object you pass
 * in.  However, if the locale data for the locale you passed into the IntlProvider was not
 * correctly installed with addLocaleData, all of your translations will fall back to the default
 * (in our case English), *even if you gave IntlProvider the correct messages object for that
 * locale*.
 */

import { addLocaleData } from 'react-intl';
import Cookies from 'universal-cookie';

import arLocale from 'react-intl/locale-data/ar';
import enLocale from 'react-intl/locale-data/en';
import es419Locale from 'react-intl/locale-data/es';
import frLocale from 'react-intl/locale-data/fr';
import zhcnLocale from 'react-intl/locale-data/zh';

import COUNTRIES, { langs as countryLangs } from 'i18n-iso-countries';
import LANGUAGES, { langs as languageLangs } from '@cospired/i18n-iso-languages';

import arMessages from './messages/ar.json';
// no need to import en messages-- they are in the defaultMessage field
import es419Messages from './messages/es_419.json';
import frMessages from './messages/fr.json';
import zhcnMessages from './messages/zh_CN.json';

addLocaleData([...arLocale, ...enLocale, ...es419Locale, ...frLocale, ...zhcnLocale]);

// TODO: When we start dynamically loading translations only for the current locale, change this.
COUNTRIES.registerLocale(require('i18n-iso-countries/langs/ar.json'));
COUNTRIES.registerLocale(require('i18n-iso-countries/langs/en.json'));
COUNTRIES.registerLocale(require('i18n-iso-countries/langs/es.json'));
COUNTRIES.registerLocale(require('i18n-iso-countries/langs/fr.json'));
COUNTRIES.registerLocale(require('i18n-iso-countries/langs/zh.json'));

// TODO: When we start dynamically loading translations only for the current locale, change this.
// TODO: Also note that Arabic (ar) and Chinese (zh) are missing here.  That's because they're
// not implemented in this library.  If you read this and it's been a while, go check and see
// if that's changed!
LANGUAGES.registerLocale(require('@cospired/i18n-iso-languages/langs/en.json'));
LANGUAGES.registerLocale(require('@cospired/i18n-iso-languages/langs/es.json'));
LANGUAGES.registerLocale(require('@cospired/i18n-iso-languages/langs/fr.json'));

const messages = { // current fallback strategy is to use the first two letters of the locale code
  ar: arMessages,
  es: es419Messages,
  fr: frMessages,
  zh: zhcnMessages,
};

const cookies = new Cookies();

const getTwoLetterLanguageCode = code => code.substr(0, 2);

// Get the locale by setting priority. Skip if we don't support that language.
const getLocale = (localeStr) => {
  // 1. Explicit application request
  if (localeStr && messages[localeStr] !== undefined) {
    return localeStr;
  }
  // 2. User setting in cookie
  const cookieLangPref = cookies.get(process.env.LANGUAGE_PREFERENCE_COOKIE_NAME);
  if (cookieLangPref && messages[getTwoLetterLanguageCode(cookieLangPref)] !== undefined) {
    return getTwoLetterLanguageCode(cookieLangPref);
  }
  // 3. Browser language (default)
  return getTwoLetterLanguageCode(window.navigator.language);
};

const getMessages = (locale = getLocale()) => messages[locale];

const rtlLocales = ['ar', 'he', 'fa', 'ur'];
const isRtl = locale => rtlLocales.includes(locale);

const handleRtl = () => {
  if (isRtl(getLocale())) {
    document.getElementsByTagName('html')[0].setAttribute('dir', 'rtl');
    document.styleSheets[0].disabled = true;
  } else {
    document.styleSheets[1].disabled = true;
  }
};

/**
 * Provides a lookup table of country IDs to country names for the current locale.
 */
const getCountryMessages = (locale) => {
  const finalLocale = countryLangs().includes(locale) ? locale : 'en';

  return COUNTRIES.getNames(finalLocale);
};

/**
 * Provides a lookup table of language IDs to language names for the current locale.
 */
const getLanguageMessages = (locale) => {
  const finalLocale = languageLangs().includes(locale) ? locale : 'en';

  return LANGUAGES.getNames(finalLocale);
};

const sortFunction = (a, b) => {
  // If localeCompare exists, use that.  (Not supported in some older browsers)
  if (typeof String.prototype.localeCompare === 'function') {
    return a[1].localeCompare(b[1], getLocale());
  }
  if (a[1] === b[1]) {
    return 0;
  }
  // Otherwise make a best effort.
  return a[1] > b[1] ? 1 : -1;
};

/**
 * Provides a list of countries represented as objects of the following shape:
 *
 * {
 *   key, // The ID of the country
 *   name // The localized name of the country
 * }
 *
 * The list is sorted alphabetically in the current locale.
 * This is useful for select dropdowns primarily.
 */
const getCountryList = (locale) => {
  const countryMessages = getCountryMessages(locale);
  return Object.entries(countryMessages)
    .sort(sortFunction)
    .map(([code, name]) => ({ code, name }));
};

/**
 * Provides a list of languages represented as objects of the following shape:
 *
 * {
 *   key, // The ID of the language
 *   name // The localized name of the language
 * }
 *
 * The list is sorted alphabetically in the current locale.
 * This is useful for select dropdowns primarily.
 */
const getLanguageList = (locale) => {
  const languageMessages = getLanguageMessages(locale);
  return Object.entries(languageMessages)
    .sort(sortFunction)
    .map(([code, name]) => ({ code, name }));
};

export {
  getCountryList,
  getCountryMessages,
  getLanguageList,
  getLanguageMessages,
  getLocale,
  getMessages,
  handleRtl,
  isRtl,
};
