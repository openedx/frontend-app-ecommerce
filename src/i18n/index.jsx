import arMessages from './messages/ar.json';
import caMessages from './messages/ca.json';
// no need to import en messages-- they are in the defaultMessage field
import es419Messages from './messages/es_419.json';
import frMessages from './messages/fr.json';
import zhcnMessages from './messages/zh_CN.json';
import heMessages from './messages/he.json';
import idMessages from './messages/id.json';
import kokrMessages from './messages/ko_kr.json';
import plMessages from './messages/pl.json';
import ptbrMessages from './messages/pt_br.json';
import ruMessages from './messages/ru.json';
import thMessages from './messages/th.json';
import ukMessages from './messages/uk.json';

const messages = {
  ar: arMessages,
  'es-419': es419Messages,
  fr: frMessages,
  'zh-cn': zhcnMessages,
  ca: caMessages,
  he: heMessages,
  id: idMessages,
  'ko-kr': kokrMessages,
  pl: plMessages,
  'pt-br': ptbrMessages,
  ru: ruMessages,
  th: thMessages,
  uk: ukMessages,
};

export default messages;
