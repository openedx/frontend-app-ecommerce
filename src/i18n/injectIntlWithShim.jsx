import React from 'react';
import { injectIntl, intlShape } from 'react-intl';
import { logError } from '@edx/frontend-logging';


const injectIntlWithShim = (WrappedComponent) => {
  class ShimmedIntlComponent extends React.Component {
    static propTypes = {
      intl: intlShape.isRequired,
    };

    constructor(props) {
      super(props);
      this.shimmedIntl = Object.create(this.props.intl, {
        formatMessage: {
          value: (definition) => {
            if (definition === undefined || definition.id === undefined) {
              const error = new Error('i18n error: An undefined message was supplied to intl.formatMessage.');
              if (process.env.NODE_ENV !== 'production') {
                console.error(error); // eslint-disable-line no-console
                return '!!! Missing message supplied to intl.formatMessage !!!';
              }
              logError(error);
              return ''; // Fail silent in production
            }
            return this.props.intl.formatMessage(definition);
          },
        },
      });
    }

    render() {
      return <WrappedComponent {...this.props} intl={this.shimmedIntl} />;
    }
  }

  return injectIntl(ShimmedIntlComponent);
};


export default injectIntlWithShim;
