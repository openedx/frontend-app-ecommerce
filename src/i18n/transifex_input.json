{
  "siteheader.links.courses": "Courses",
  "siteheader.links.programs": "Programs & Degrees",
  "siteheader.links.schools": "Schools & Partners",
  "siteheader.user.menu.dashboard": "Dashboard",
  "siteheader.user.menu.profile": "Profile",
  "siteheader.user.menu.account.settings": "Account",
  "siteheader.user.menu.logout": "Logout",
  "siteheader.user.menu.login": "Login",
  "siteheader.user.menu.register": "Sign Up",
  "error.unexpected.message": "An unexpected error occurred.",
  "error.notfound.message": "The page you're looking for is unavailable or there's an error in the URL. Please check the URL and try again.",
  "root.page": "Congratulations!  You have a new micro-frontend.",
  "ecommerce.order.history.page.heading": "Order History",
  "ecommerce.order.history.no.orders": "Orders you place with edX will appear here.",
  "ecommerce.order.history.loading.orders": "Loading orders...",
  "ecommerce.order.history.loading.error": "Error: {error}"
}