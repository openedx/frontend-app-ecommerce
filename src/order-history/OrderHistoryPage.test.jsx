/* eslint-disable global-require */
import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { IntlProvider } from '@edx/frontend-platform/i18n';
import configureMockStore from 'redux-mock-store';

import ConnectedOrderHistoryPage from './OrderHistoryPage';

const mockStore = configureMockStore();
const storeMocks = {
  ordersLoaded: require('./__mocks__/ordersLoaded.mockStore'),
};
const requiredOrderHistoryPageProps = {
  isB2CSubsEnabled: false,
  fetchOrders: () => {},
};

// Match all media queries. This will result in rendering
// both the desktop and mobile views at the same time.
global.matchMedia = media => ({ // eslint-disable-line no-unused-vars
  addListener: () => {},
  removeListener: () => {},
  matches: true,
});

describe('<OrderHistoryPage />', () => {
  describe('Renders correctly in various states', () => {
    it('renders orders table with pagination', () => {
      const tree = renderer
        .create((
          <IntlProvider locale="en">
            <Provider store={mockStore(storeMocks.ordersLoaded)}>
              <ConnectedOrderHistoryPage {...requiredOrderHistoryPageProps} />
            </Provider>
          </IntlProvider>
        ))
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
