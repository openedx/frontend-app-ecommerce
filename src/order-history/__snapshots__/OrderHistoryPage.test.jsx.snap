// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OrderHistoryPage /> Renders correctly in various states renders orders table with pagination 1`] = `
<div
  className="page__order-history container-fluid py-5"
>
  <h1>
    Order History
  </h1>
  <nav
    aria-label="pagination navigation"
    className="pagination-default"
  >
    <div
      aria-atomic={true}
      aria-live="polite"
      aria-relevant="text"
      className="sr-only"
    >
      Page 2, Current Page, of 10
    </div>
    <ul
      className="pagination"
    >
      <li
        className="page-item"
      >
        <button
          aria-label="Previous, Page 1"
          className="previous page-link btn btn-primary"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <div>
            <span
              className="pgn__icon"
            >
              <svg
                aria-hidden={true}
                fill="none"
                focusable={false}
                height={24}
                role="img"
                viewBox="0 0 24 24"
                width={24}
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12l4.58-4.59z"
                  fill="currentColor"
                />
              </svg>
            </span>
            Previous
          </div>
        </button>
      </li>
      <li
        className="page-item"
      >
        <button
          aria-label="Next, Page 3"
          className="next page-link btn btn-primary"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <div>
            Next
            <span
              className="pgn__icon"
            >
              <svg
                aria-hidden={true}
                fill="none"
                focusable={false}
                height={24}
                role="img"
                viewBox="0 0 24 24"
                width={24}
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6-6-6z"
                  fill="currentColor"
                />
              </svg>
            </span>
          </div>
        </button>
      </li>
    </ul>
  </nav>
</div>
`;
